tools/plugin/alsaplug/tplg_ctl.c

```diff
@@ -0,0 +1,204 @@
+	if (exp < 0)
+		iter = exp * -1;
+	else
+		iter = exp;
+
+	/* multiply a "iter" times to compute power */
+	for (i = 0; i < iter; i++) {
+		/*
+		 * Product of 2 Qx.fwl fixed-point numbers yields a Q2*x.2*fwl
+		 * Truncate product back to fwl fractional bits with rounding
+		 */
+		power = vol_shift_64((uint64_t)power * a, fwl);
+	}
+
+	if (exp > 0) {
+		/* if exp is positive, return the result */
+		return power;
+	}
+
+	/* if exp is negative, return the multiplicative inverse */
+	numerator = (uint64_t)1 << (fwl << 1);
+	numerator /= power;
+
+	return (uint32_t)numerator;
+}
+
+/*
+ * Function to calculate volume gain from TLV data.
+ * This function can only handle gain steps that are multiples of 0.5 dB
+ */
+static uint32_t vol_compute_gain(uint32_t value, struct snd_soc_tplg_tlv_dbscale *scale)
+{
+	int dB_gain;
+	uint32_t linear_gain;
+	int f_step;
+
+	/* mute volume */
+	if (value == 0 && scale->mute)
+		return 0;
+
+	/*
+	 * compute dB gain from tlv. tlv_step
+	 * in topology is multiplied by 100
+	 */
+	dB_gain = (int)scale->min / 100 + (value * scale->step) / 100;
+	/*
+	 * compute linear gain represented by fixed-point
+	 * int with VOLUME_FWL fractional bits
+	 */
+	linear_gain = vol_pow32(VOL_TWENTIETH_ROOT_OF_TEN, dB_gain, VOLUME_FWL);
+
+	/* extract the fractional part of volume step */
+	f_step = scale->step - (scale->step / 100);
+
+	/* if volume step is an odd multiple of 0.5 dB */
+	if (f_step == VOL_HALF_DB_STEP && (value & 1))
+		linear_gain = vol_shift_64((uint64_t)linear_gain * VOL_FORTIETH_ROOT_OF_TEN,
+					   VOLUME_FWL);
+
+	return linear_gain;
+}
+
+/* helper function to add new kcontrols to the list of kcontrols in the global context */
+int plug_kcontrol_cb_new(struct snd_soc_tplg_ctl_hdr *tplg_ctl, void *_comp, void *arg)
+{
+	snd_sof_plug_t *plug = arg;
+	struct plug_shm_glb_state *glb = plug->glb_ctx.addr;
+	struct plug_shm_ctl *ctl;
+
+	if (glb->num_ctls >= MAX_CTLS) {
+		SNDERR("Failed to add a new control. Too many controls already\n");
+		return -EINVAL;
+	}
+
+	switch (tplg_ctl->type) {
+	case SND_SOC_TPLG_CTL_VOLSW:
+	case SND_SOC_TPLG_CTL_VOLSW_SX:
+	case SND_SOC_TPLG_CTL_VOLSW_XR_SX:
+	{
+		struct snd_soc_tplg_mixer_control *tplg_mixer =
+			(struct snd_soc_tplg_mixer_control *)tplg_ctl;
+		struct tplg_comp_info *comp_info = _comp;
+		struct snd_soc_tplg_ctl_tlv *tlv;
+		struct snd_soc_tplg_tlv_dbscale *scale;
+		int i;
+
+		glb->size += sizeof(struct plug_shm_ctl);
+		ctl = &glb->ctl[glb->num_ctls++];
+		ctl->module_id = comp_info->module_id;
+		ctl->instance_id = comp_info->instance_id;
+		ctl->mixer_ctl = *tplg_mixer;
+		tlv = &tplg_ctl->tlv;
+		scale = &tlv->scale;
+
+		/* populate the volume table */
+		for (i = 0; i < tplg_mixer->max + 1 ; i++) {
+			uint32_t val = vol_compute_gain(i, scale);
+
+			/* Can be over Q1.31, need to saturate */
+			uint64_t q31val = ((uint64_t)val) << 15;
+
+			ctl->volume_table[i] = q31val > SOF_IPC4_VOL_ZERO_DB ?
+							SOF_IPC4_VOL_ZERO_DB : q31val;
```

Comment:
is some kind of a `MIN()` macro available?

Reply:
This code is copied as is from the linux driver. Is it OK to leave it as is for the sake of consistency?